lambdas/select-cri/src/main/java/uk/gov/di/ipv/core/selectcri/SelectCriHandler.java

```diff
@@ -217,8 +246,94 @@ private APIGatewayProxyResponseEvent getJourneyResponse(String criId) {
                 HttpStatus.SC_OK, new JourneyResponse(String.format(CRI_START_JOURNEY, criId)));
     }
 
+    private APIGatewayProxyResponseEvent getJourneyPyiNoMatchResponse() {
+        return ApiGatewayResponseGenerator.proxyJsonResponse(
+                HttpStatus.SC_OK, new JourneyResponse("/journey/pyi-no-match"));
+    }
+
     private boolean userHasNotVisited(
             List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers, String criId) {
         return visitedCredentialIssuers.stream().noneMatch(cri -> cri.getCriId().equals(criId));
     }
+
+    private Optional<APIGatewayProxyResponseEvent> getCriResponse(
+            List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers,
+            BiPredicate<CredentialEvidenceItem, Gpg45Profile> gpg45Validator,
+            Gpg45Profile gpg45Profile,
+            String criId,
+            String userId)
+            throws ParseException {
+        if (userHasNotVisited(visitedCredentialIssuers, criId)) {
+            return Optional.of(getJourneyResponse(criId));
+        } else {
+            Optional<APIGatewayProxyResponseEvent> failedJourneyResponse =
+                    getFailedJourneyResponse(
+                            visitedCredentialIssuers, gpg45Validator, gpg45Profile, criId, userId);
+
+            if (failedJourneyResponse.isPresent() && criId.equals(dcmawCriId)) {
+                return Optional.of(getNextWebJourneyCri(visitedCredentialIssuers, userId));
+            }
+            return failedJourneyResponse;
+        }
+    }
+
+    private Optional<APIGatewayProxyResponseEvent> getFailedJourneyResponse(
+            List<VisitedCredentialIssuerDetailsDto> visitedCredentialIssuers,
+            BiPredicate<CredentialEvidenceItem, Gpg45Profile> validator,
+            Gpg45Profile gpg45Profile,
+            String criId,
+            String userId)
+            throws ParseException {
+        Optional<VisitedCredentialIssuerDetailsDto> criVisitDetails =
+                visitedCredentialIssuers.stream()
+                        .filter(cri -> cri.getCriId().equals(criId))
+                        .findFirst();
+
+        if (criVisitDetails.isPresent()) {
+            if (criVisitDetails.get().isReturnedWithVc()) {
+                if (!isSuccessfulVc(userId, criId, validator, gpg45Profile)) {
+                    LOGGER.info(
+                            "User has a previous failed visit to {} cri due to a failed identity check. Routing user to failed journey path",
+                            criId);
+                    return Optional.of(getJourneyPyiNoMatchResponse());
+                }
+            } else {
+                LOGGER.info(
+                        "User has a previous failed visit to {} cri due to: {}. Routing user to web journey instead.",
```

Comment:
Just realised this log isn't quite right anymore. I'll move this one and the one above slightly so that it can reflect correctly what actually happens